File: src/dotnet/Ansight.Tools.Reflection/SetMemberValueTool.cs
namespace Ansight.Tools.Reflection;

public sealed class SetMemberValueTool : ITool
{
    private readonly ReflectionToolsOptions options;

    public SetMemberValueTool(ReflectionToolsOptions? options = null)
    {
        this.options = options ?? ReflectionToolsOptions.Default;
    }

    public string Category => "reflect";

    public ToolScope Scope => ToolScope.Write;

    public string Id => ReflectionToolIds.SetMemberValue;

    public string Name => "Set Member Value";

    public string Description => "Writes a writable field or property reachable from a registered live object root.";

    public string Keywords => "reflection set write property field runtime";

    public ToolSchema ArgumentsSchema => ReflectionToolSchemas.SetMemberValueArguments;

    public ToolSchema ResultSchema => ReflectionToolSchemas.SetMemberValueResult;

    public ToolSecurity Security => ReflectionToolSecurityProfiles.SetMemberValue;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            return Task.FromResult(ToolResult.Success(ReflectionSupport.SetMemberValue(options, arguments)));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "reflect_set_member_failed"));
        }
    }
}

File: src/ios/Sources/AnsightToolsReflection/SetMemberValueTool.swift
import AnsightCore
import Foundation

public final class SetMemberValueTool: AnsightTool {
    private let runtime: AnsightRuntime
    private let options: AnsightReflectionToolsOptions

    public init(
        options: AnsightReflectionToolsOptions = .default,
        runtime: AnsightRuntime = .shared
    ) {
        self.runtime = runtime
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightReflectionToolIds.setMemberValue,
            name: "Set Member Value",
            description: "Writes an opt-in member reachable from a registered iOS object root.",
            category: "reflect",
            scope: AnsightToolScope.write.rawValue,
            keywords: "reflection set write property field runtime",
            security: AnsightReflectionToolSecurityProfiles.setMemberValue,
            argumentsSchema: AnsightReflectionToolSchemas.setMemberValueArguments,
            resultSchema: AnsightReflectionToolSchemas.setMemberValueResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            return .success(try AnsightReflectionSupport.setMemberValue(runtime: runtime, options: options, arguments: arguments))
        } catch {
            return .failure(error.localizedDescription, errorCode: "reflect_set_member_failed")
        }
    }
}
